stm32: Support for SDMMC on STM32H7

fzawadiak opened the pull request:
Extend STM32 SDMMC driver to work with STM32H7 CPUs.
Change configuration of STM32H747I-DISCO to enable it.

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

erwango commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -37,6 +37,21 @@ config CLOCK_STM32_D3PPRE
+	default 4
+	
+config CLOCK_STM32_PLL2_R_DIVISOR
+	default 5
```
A comment explaining which speed you expect with this setting would be nice.

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -64,6 +64,32 @@ static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
 		;
 
 	LL_RCC_SetSDMMCClockSource(LL_RCC_SDMMC1_CLKSOURCE_PLLSAI1);
+#elseif CONFIG_SOC_SERIES_STM32H7X
```
```suggestion
#elif defined(CONFIG_SOC_SERIES_STM32H7X)
```

fzawadiak replied:
Done

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -64,6 +64,32 @@ static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
+	LL_RCC_PLL2_SetQ(CONFIG_CLOCK_STM32_PLL2_Q_DIVISOR);
+	LL_RCC_PLL2_SetR(CONFIG_CLOCK_STM32_PLL2_R_DIVISOR);
+	
+	LL_RCC_PLL2_Enable();
```
Move this part to `stm32_clock_control_init` in `drivers/clock_control/clock_stm32_ll_h7.c`

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -64,6 +64,32 @@ static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
+	
+	LL_RCC_PLL2_Enable();
+	
+	LL_RCC_SetSDMMCClockSource(LL_RCC_SDMMC_CLKSOURCE_PLL2R);
```
Since PLL configuration is done at another place, check that right configuration is enabled (CLOCK_STM32_PLL2_ENABLE)

erwango reviewed (changes requested):
@fzawadiak Thanks for this PR. I don't think there is too much to change.
See comments.

Aside from pure code changes, one thing to modify is the format of your PR:
- We don't merge fix commits, so you need to squash them
- On the other hand, you need to split your PR in topics commits (boards, dts, one per driver, ..)

fzawadiak committed:
drivers/clock_control: stm32: Support for PLL2 configuration in STM32H7

PLL2 can be configured using Kconfig in STM32H7.
Code synced with PR 30906

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak committed:
subsys/disk: stm32: Added support for STM32H7 in SDMMC driver

SDMMC driver supports STM32H7 in interrupt mode with DMA.

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak committed:
dts: stm32: Support for SDMMC driver in STM32H7

Entries for SDMMC added in STM32H7 DT plus change in
bindings to enable interrupts.

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak committed:
boards: stm32h747i-disco: Configuration of SDMMC to STM32H747I-DISCO

Board can be configured for SDMMC

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak committed:
doc: stm32: Update of STM32H747I-DISCO board about SDMMC support

Updated documentation of STM32H747I-DISCO about support of SDMMC

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

gmarull commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -675,6 +675,32 @@ static int stm32_clock_control_init(const struct device *dev)
+	LL_RCC_PLL2FRACN_Disable();
+	LL_RCC_PLL2P_Enable();
+	LL_RCC_PLL2Q_Enable();
+	LL_RCC_PLL2R_Enable();
```
It should be possible to enable channels individually, if I'm not wrong SDMMC only needs R if using PLL2. I have pushed support PLL3 in https://github.com/zephyrproject-rtos/zephyr/pull/30906, it may be useful

fzawadiak replied:
So possibly all PLLs should be configurable the same way with Kconfig. Then it makes sense to start adding clock source selectors for all drivers? So far I have not found it in any...

erwango replied:
> So possibly all PLLs should be configurable the same way with Kconfig. Then it makes sense to start adding clock source selectors for all drivers? So far I have not found it in any...

This is indeed something that will be requested, but it doesn't make much sense to add it right now using Kconfig. Let's live without it few months more and get this using dts.

gmarull commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -675,6 +675,32 @@ static int stm32_clock_control_init(const struct device *dev)
+	/* Config PLL */
+	/* VCO sel, VCO range */
+	LL_RCC_PLL2_SetVCOInputRange(LL_RCC_PLLINPUTRANGE_1_2);
+	LL_RCC_PLL2_SetVCOOutputRange(LL_RCC_PLLVCORANGE_WIDE);
```
Input/Output range should be calculated dynamically

fzawadiak replied:
Thanks, I did it like you did for PLL3

gmarull commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -64,6 +64,8 @@ static int stm32_sdmmc_clock_enable(struct stm32_sdmmc_priv *priv)
 
 	LL_RCC_SetSDMMCClockSource(LL_RCC_SDMMC1_CLKSOURCE_PLLSAI1);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+	LL_RCC_SetSDMMCClockSource(LL_RCC_SDMMC_CLKSOURCE_PLL2R);
```
PLL1Q or PLL2R can be selected, so it should probably be configurable? 

![image](https://user-images.githubusercontent.com/25011557/102665712-97502580-4185-11eb-998c-446d6160dfc0.png)

fzawadiak replied:
Yes, there are many ways to configure this clock. Don't know whether this is good moment to introduce too many options, it is so complex it probably would make sense to introduce PLL child object under RCC and give clock source selection for each driver. Good time will be after SDMMC and QSPI work together.

fzawadiak replied:
I added option to select clock source, might be already useful

gmarull commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -724,6 +724,14 @@
 			};
 		};
+
+		sdmmc1: sdmmc@52007000 {
```
H7 series have `sdmmc1` and `sdmmc2`, I think both should be added.

fzawadiak replied:
I can't test the other one, prefer to add things I got working.

fzawadiak replied:
Added

gmarull reviewed (changes requested):
Thanks for this addition, left a few comments

gmarull reviewed (commented):
Thanks @fzawadiak. LGTM, regarding PLL it's probably good to synchronize the changes, I think 17dfcc9c8917c57cffe545973ff7d3fe4fc5a864 from https://github.com/zephyrproject-rtos/zephyr/pull/30906 should go first, then any PLL (2 or 3) can be added. You may cherry-pick it and add PLL2 support on top.

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -298,12 +298,28 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
 	defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
 
-static int32_t get_vco_output_range(uint32_t vco_input_range)
+static int get_vco_input_range(uint32_t m_div, uint32_t *range)
+{
+	uint32_t vco_freq;
+	vco_freq = PLLSRC_FREQ / m_div;
+	if (1000000UL <= vco_freq && vco_freq <= 2000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_1_2;
+	} else if (2000000UL < vco_freq && vco_freq <= 4000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_2_4;
+	} else if (4000000UL < vco_freq && vco_freq <= 8000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_4_8;
+	} else if (8000000UL < vco_freq && vco_freq <= 16000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_8_16;
+	} else {
+		return -ERANGE;
+	}
+	return 0;
+}
```
I do not see the need of `get_vco_input_range()` function definition, since the calculations are done with preprocessor on the lines  69-81 of this file.
New line is missed also.

gmarull replied:
@Nukersson It is needed when using multiple PLLs.

Nukersson replied:
Okay, then another proposal. Please put defines in the top of the file. This defines let programmer read code easier, cause no zeros counting is needed.

```suggestion
#define FREQ_1MHz     1000000UL
#define FREQ_2MHz      2000000UL
#define FREQ_4MHz      4000000UL
#define FREQ_8MHz      8000000UL
#define FREQ_16MHz    16000000UL

static int get_vco_input_range(uint32_t m_div, uint32_t *range)
{
	uint32_t vco_freq;
	vco_freq = PLLSRC_FREQ / m_div;
	if (FREQ_1MHz <= vco_freq && vco_freq <= FREQ_2MHz) {
		*range = LL_RCC_PLLINPUTRANGE_1_2;
	} else if (FREQ_2MHz < vco_freq && vco_freq <= FREQ_4MHz) {
		*range = LL_RCC_PLLINPUTRANGE_2_4;
	} else if (4000000UL < vco_freq && vco_freq <= FREQ_8MHz) {
		*range = LL_RCC_PLLINPUTRANGE_4_8;
	} else if (8000000UL < vco_freq && vco_freq <= FREQ_16MHz) {
		*range = LL_RCC_PLLINPUTRANGE_8_16;
	} else {
		return -ERANGE;
	}
	return 0;
}

```

fzawadiak replied:
Would it not make more sense to have one define like FREQ_MHZ(x) (x*1000000UL)?

Nukersson replied:
> Would it not make more sense to have one define like FREQ_MHZ(x) (x*1000000UL)?

sure

fzawadiak replied:
Done

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -670,6 +692,35 @@ static int stm32_clock_control_init(const struct device *dev)
+#endif /* CONFIG_CLOCK_STM32_PLL2_R_ENABLE */
+	LL_RCC_PLL2_Enable();
+	while (LL_RCC_PLL2_IsReady() != 1) {
+	}
+#endif /* CONFIG_CLOCK_STM32_PLL2_ENABLE */
```
```suggestion
	}
#else
	/* PLL will stay in reset state configuration */
#endif /* CONFIG_CLOCK_STM32_PLL2_ENABLE */
```

`#else` is not necessary, bur provides same way definition as for PLL1 then.

fzawadiak replied:
Done

KozhinovAlexander reviewed (changes requested):
Thank your for your PR. It is very welcome change in Zephyr! Please integrate change-requests or tell the reason not to use them.

gmarull commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.dts:
```diff
@@ -127,3 +127,10 @@
+
+&sdmmc1 {
+	status = "okay";
+	pinctrl-0 = <&sdmmc1_d0_pc8 &sdmmc1_d1_pc9 &sdmmc1_d2_pc10 &sdmmc1_d3_pc11
+				&sdmmc1_ck_pc12 &sdmmc1_cmd_pd2>;
```
@Nukersson Note that other pinctrl do not follow one entry per line, so current format should be OK

erwango replied:
I agree the more dense version is OK. I even prefer this one (with correct indentation maybe) to enjoy shorter files.

fzawadiak replied:
OK, will fix this. All other suggestions I did, now figuring out cherry picking process :-)

Nukersson replied:
> @Nukersson Note that other pinctrl do not follow one entry per line, so current format should be OK

I am not a fon of smaller lines number here, cause Reference Manuals have always table for pins definition. Therefore it is usually easier to go through pins in some kind of table especially if you doing it parallel in RM and code.

fzawadiak replied:
I fixed (?) indentation. It did not look very broken though...

fzawadiak commented:
I add code to enable SDMMC2 in subsys/disk/disk_access_stm32_sdmmc.c

KozhinovAlexander reviewed (approved):
LGTM

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -439,3 +445,42 @@ DEVICE_DT_INST_DEFINE(0, disk_stm32_sdmmc_init, device_pm_control_nop,
 #endif
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+
+static const struct soc_gpio_pinctrl sdmmc_pins_2[] =
+	ST_STM32_DT_INST_PINCTRL(1, 0);
+
+static struct stm32_sdmmc_priv stm32_sdmmc_priv_2 = {
+	.hsd = {
+		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(1),
+	},
+#if DT_INST_NODE_HAS_PROP(1, cd_gpios)
+	.cd = {
+		.name = DT_INST_GPIO_LABEL(1, cd_gpios),
+		.pin = DT_INST_GPIO_PIN(1, cd_gpios),
+		.flags = DT_INST_GPIO_FLAGS(1, cd_gpios),
+	},
+#endif
+#if DT_INST_NODE_HAS_PROP(1, pwr_gpios)
+	.pe = {
+		.name = DT_INST_GPIO_LABEL(1, pwr_gpios),
+		.pin = DT_INST_GPIO_PIN(1, pwr_gpios),
+		.flags = DT_INST_GPIO_FLAGS(1, pwr_gpios),
+	},
+#endif
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(1, bus),
+		.enr = DT_INST_CLOCKS_CELL(1, bits),
+	},
+	.pinctrl = {
+		.list = sdmmc_pins_2,
+		.len = ARRAY_SIZE(sdmmc_pins_2)
+	}
+};
+
+DEVICE_DT_INST_DEFINE(1, disk_stm32_sdmmc_init, device_pm_control_nop,
+		    &stm32_sdmmc_priv_2, NULL, APPLICATION,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    NULL);
+#endif
```
Please use an instantiation macro, similarly to what is done for other multi instance drivers. 
For instance:
https://github.com/zephyrproject-rtos/zephyr/blob/5d7fc17034b29087af801577baadca7f96b64fd4/drivers/spi/spi_ll_stm32.c#L918-L950

fzawadiak replied:
Tried this in the beginning, but gave up on conditional cd and pwr definitions...
Something like this would work, but gives few pages of errors when compiling. Any hints where I can find something similar as reference?
```c
#define STM32_SDMMC_GPIO_INIT(id, func)					\
	.##func = {							\
		.name = DT_INST_GPIO_LABEL(id, func##_gpios),		\
		.pin = DT_INST_GPIO_PIN(id, func##_gpios),		\
		.flags = DT_INST_GPIO_FLAGS(id, func##_gpios),		\
	}

#define STM32_SDMMC_GPIO(id, func)					\
	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, func##_gpios),		\
		(STM32_SDMMC_GPIO_INIT(id, func)),			\
		(NULL))							\
	}

#define STM32_SDMMC_INIT(id)						\
									\
static const struct soc_gpio_pinctrl sdmmc_pins_##id[] =		\
	ST_STM32_DT_INST_PINCTRL(id, 0);				\
									\
static struct stm32_sdmmc_priv stm32_sdmmc_priv_##id = {		\
	.hsd = {							\
		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(id),	\
	},								\
	.pclken = {							\
		.bus = DT_INST_CLOCKS_CELL(id, bus),			\
		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
	},								\
	.pinctrl = {							\
		.list = sdmmc_pins_##id,				\
		.len = ARRAY_SIZE(sdmmc_pins_##id)			\
	},								\
	STM32_SDMMC_GPIO(id, cd),					\
	STM32_SDMMC_GPIO(id, pwr),					\
};									\
									\
DEVICE_DT_INST_DEFINE(id, disk_stm32_sdmmc_init, device_pm_control_nop,	\
		    &stm32_sdmmc_priv_##id, NULL, APPLICATION,		\
		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
		    NULL)

DT_INST_FOREACH_STATUS_OKAY(STM32_SDMMC_INIT)
```

erwango replied:
I got the following working:
```
#define STM32_SDMMC_GPIO_CONFIG(id, func)				\
		.name = DT_INST_GPIO_LABEL(id, func##_gpios),		\
		.pin = DT_INST_GPIO_PIN(id, func##_gpios),		\
		.flags = DT_INST_GPIO_FLAGS(id, func##_gpios),		\

#define STM32_SDMMC_GPIO(id, func)					\
	.func = {							\
		COND_CODE_1(DT_INST_NODE_HAS_PROP(id, func##_gpios),	\
		(STM32_SDMMC_GPIO_CONFIG(id, func)),			\
		(NULL))							\
	}

#define STM32_SDMMC_INIT(id)						\
									\
static const struct soc_gpio_pinctrl sdmmc_pins_##id[] =		\
	ST_STM32_DT_INST_PINCTRL(id, 0);				\
									\
static struct stm32_sdmmc_priv stm32_sdmmc_priv_##id = {		\
	.hsd = {							\
		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(id),	\
	},								\
	.pclken = {							\
		.bus = DT_INST_CLOCKS_CELL(id, bus),			\
		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
	},								\
	.pinctrl = {							\
		.list = sdmmc_pins_##id,				\
		.len = ARRAY_SIZE(sdmmc_pins_##id)			\
	},								\
	STM32_SDMMC_GPIO(id, cd),					\
	STM32_SDMMC_GPIO(id, pe),					\
};									\
									\
DEVICE_DT_INST_DEFINE(id, disk_stm32_sdmmc_init, device_pm_control_nop,	\
		    &stm32_sdmmc_priv_##id, NULL, APPLICATION,		\
		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
		    NULL);						\

DT_INST_FOREACH_STATUS_OKAY(STM32_SDMMC_INIT)

```

Though, I had to change `STM32_SDMMC_GPIO(id, pwr),` to `STM32_SDMMC_GPIO(id, pe),`.
Can you check this particular point ?

fzawadiak replied:
Won't this set .func first value to null instead of whole null in case pin is not defined?
I think having defines repeated twice makes sense - too much preprocessor magic for me otherwise...

erwango replied:
> Won't this set .func first value to null instead of whole func in case pin is not defined?

This was a quick hack, so I can't swear it's bug free, though this kind of processing is used at other places in tree.

> I think having defines repeated twice makes sense - too much preprocessor magic for me otherwise...

I can understand this feels a lot of preprocessor treatment for a little gain.
But this is the way it is done in other drivers so it makes sense to make it consistent with other drivers. One gain is the minimized maintenance vs both instances. 
Anyway I won't block for this... but I will make sure it is done afterwards.

fzawadiak replied:
Lets have a loop then, I'm fine with pe too. Possibly it is used by other devices and I did not define pwr properly before?

erwango commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7_defconfig:
```diff
@@ -43,3 +43,6 @@ CONFIG_SERIAL=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+
+# enable SDMMC driver
+CONFIG_DISK_ACCESS_STM32_SDMMC=y
```
Please remove this.
Flag will be enabled depending on sdmmc1 node activation and DISK_ACCESS activation by application

fzawadiak replied:
But then driver source files is not included. Should I really remove it?

erwango replied:
This symbol will be set to 'y' if a node with "st,stm32-sdmmd" is set to `status = "okay";` and `CONFIG_DISK_ACCESS=y`.
So, it will be enabled when needed (ie requested by application).
We're doing the same will all other peripheral symbols (aside from those always enabled because always requested: SERIAL, CLOCK_CONTROL, INTERRUPT_CONTROLLER, PINMUX).

cf https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations

erwango replied:
Please consider this one as the last blocking comment.

fzawadiak replied:
OK, CONFIG_DISK_ACCESS removed, code synced with PLL3 changes

gmarull commented:
@fzawadiak PLL3 changes have been merged, so you should be able to rebase this PR on top.

fzawadiak commented:
Super, will update my PR soon

gmarull commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -38,6 +38,27 @@ config CLOCK_STM32_D3PPRE
+config CLOCK_STM32_PLL2_N_MULTIPLIER
+	default 100
+
+config CLOCK_STM32_PLL2_P_ENABLE
```
why is `_P` channel required?

fzawadiak replied:
It was easier to set transfer speed separately

fzawadiak replied:
Will check whether it works with other channel

gmarull replied:
how are PLL settings set now?

gmarull replied:
is `_P` required? seems like `_R` is only needed, which is auto selected

fzawadiak replied:
Right, not needed. Disabled 👍

gmarull commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -38,6 +38,27 @@ config CLOCK_STM32_D3PPRE
+config DISK_ACCESS_STM32_SDMMC
+	default n
+
+config DISK_ACCESS_STM32_CLOCK_SOURCE_PLL2_R
```
Afaik choices are usually selected like this:
```
choice DISK_ACCESS_STM32_CLOCK_SOURCE
	default DISK_ACCESS_STM32_CLOCK_SOURCE_PLL2_R
endchoice
```

Example:
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/shields/buydisplay_2_8_tft_touch_arduino/Kconfig.defconfig#L48

fzawadiak replied:
Will change it asap

fzawadiak replied:
If I remove this setting as requested I can't get SDMMC driver to compile... Do I need to add some conditions based on DISK_ACCESS manually?

fzawadiak replied:
OK, got it to work with changed rule in Kconfig, like this:
```
if DISK_ACCESS

config DISK_ACCESS_STM32_SDMMC
	default y

endif # DISK_ACCESS
```

fzawadiak replied:
OK, changed the choice as recommended

fzawadiak commented:
Hmm, after recent changes it doesn't work anymore :-(
Need to do in depth debugging what was changed.

fzawadiak commented:
OK, fixed. Works with samples/subsys/fs/fat_fs without any config changes.
I also removed all PLL2 configuration as it works without it as well...

erwango commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -38,6 +38,13 @@ config CLOCK_STM32_D3PPRE
 
+if DISK_ACCESS
+
+config DISK_ACCESS_STM32_SDMMC
```
This is not needed: this symbol is defined in `subsys/disk/Kconfig` as follows:
```
DT_COMPAT_ST_STM32_SDMMC := st,stm32-sdmmc
[...]
if DISK_ACCESS
[...]
config DISK_ACCESS_STM32_SDMMC
	bool "STM32 SDMMC driver"
	depends on HAS_STM32CUBE
	select USE_STM32_HAL_SD
	select USE_STM32_LL_SDMMC
	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_SDMMC))
	help
	  File system on sdmmc accessed through stm32 sdmmc.
[...]
endif # DISK_ACCESS
```

This means it will be set to 'y' if `CONFIG_DISK_ACCESS=y` and a node with compat `st,stm32-sdmmc` is enabled.

erwango replied:
Removing this would fix CI issue.

fzawadiak replied:
Changed, sample now works

erwango commented:
@fzawadiak Doesn't work on my side:
```
*** Booting Zephyr OS build zephyr-v2.4.0-2968-g42c832f5e87d  ***
Error mounting disk.
[00:00:00.005,000] <err> main: Storage init ERROR!
[00:00:00.005,000] <err> fs: fs mount error (-5)
```
Are you sure the PLL2 init is not requested ?

fzawadiak commented:
I put back PLL configuration, it would be unpredictable to leave it depending of whatever else is done with main PLL...

erwango commented:
Still getting issues on my board.
Error happens at following level in HAL `SD_PowerON`
```
  errorstate = SDMMC_CmdOperCond(hsd->Instance);
  if(errorstate != HAL_SD_ERROR_NONE)
  {
    hsd->SdCard.CardVersion = CARD_V1_X;
    /* CMD0: GO_IDLE_STATE */
    errorstate = SDMMC_CmdGoIdleState(hsd->Instance);
    if(errorstate != HAL_SD_ERROR_NONE)
    {
      return errorstate;
    }
```
Same SD card is working well in stm32f746g_disco. 
@fzawadiak Did you do any modification on the board (jumper...)?

fzawadiak commented:
Except for Ethernet jumpers no modifications...

erwango commented:
> Except for Ethernet jumpers no modifications... Testing with Toshiba 8GB card.

Could you have a try to perform a full chip erase and see if this is still working?
I'd like to be sure this isn't working due to an obscure configuration conflict.

erwango commented:
@Nukersson, @gmarull are you able to test this?
Not that it is mandatory to test before merging, but since it's not working in my case, I prefer to be sure the issue is on my side rather than in the PR.

Nukersson commented:
> @Nukersson, @gmarull are you able to test this?
> Not that it is mandatory to test before merging, but since it's not working in my case, I prefer to be sure the issue is on my side rather than in the PR.

I do not have a possibility to test it.

fzawadiak commented:
Did chip erase, depower and reprogram. Correct SD card size is reported. Will check on FAT card tomorrow...

fzawadiak commented:
Tried with fresh SD card. Still causes problems. I noticed that board was not flashable afterwards without holding reset during erase, what usually happens when I put clocks way too high. Will debug further after weekend.

fzawadiak commented:
Seems that driver is working ok, but filesystem can’t be mounted because get_ldnumber can’t find SD: prefix.

erwango commented:
This PR missed the V2.5.0 window. Let's target V2.6.0 (hopefully as soon as merge window is re-opened)

fzawadiak commented:
I had added interrupt driven support and cleaned up default configuration to be exactly as what works in my other rtos implementation. I will also probably add DMA mode as soon as I figure out what is wrong with fat_fs example - it still gives error message...

fzawadiak committed:
samples: subsys/disk: SD card RW checker

This sample writes data to 100 sectors of SD and then verifies readout.
Useful to test new disk access API implementations and to observe
differences between pooling, interrupt and DMA modes.

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak committed:
dts: stm32: Interrupt configuration for SDMMC in STM32

Added interrupts fields for STM32F7, STM32L4 to be compatible with
improved SDMMC driver on STM32H7.

Signed-off-by: Filip Zawadiak <fzawadiak@gmail.com>

fzawadiak commented:
I'm sick of debugging FatFs code. I added new sample in samples/subsystem/diskaccess/rwcheck to prove that driver is working correctly. I also added interrupts with DMA handling, makes the whole thing fly in comparison to pooling mode... Could test only on H7.

erwango commented:
@fzawadiak Got a new stm32h747_disco and now fatfs sample is working as expected!

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -108,12 +120,38 @@ static int stm32_sdmmc_access_init(struct disk_info *disk)
+	/* Run IRQ init */
+	priv->irq_config(dev);
+
+	priv->hsd.Init.ClockEdge = SDMMC_CLOCK_EDGE_RISING;
+	priv->hsd.Init.ClockPowerSave = SDMMC_CLOCK_POWER_SAVE_DISABLE;
+	priv->hsd.Init.BusWide = SDMMC_BUS_WIDE_4B;
+	priv->hsd.Init.HardwareFlowControl = SDMMC_HARDWARE_FLOW_CONTROL_DISABLE;
+	priv->hsd.Init.ClockDiv = 1;
+	priv->hsd.Init.TranceiverPresent = SDMMC_TRANSCEIVER_NOT_PRESENT;
```
This part is breaking compilation on stm32f746g_disco

fzawadiak replied:
This and then F7 has no build-in DMA. I disabled DMA mode for anything except H7 and even there it is Kconfig option. Driver still works on H7 :-)

erwango commented on subsys/disk/Kconfig:
```diff
@@ -144,11 +144,37 @@ config DISK_ACCESS_STM32_SDMMC
 	bool "STM32 SDMMC driver"
 	depends on HAS_STM32CUBE
 	select USE_STM32_HAL_SD
+	select USE_STM32_HAL_DMA if SOC_SERIES_STM32F7X
```
Is that requested ?

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -139,15 +180,43 @@ static int stm32_sdmmc_access_read(struct disk_info *disk, uint8_t *data_buf,
-				num_sector, 30000);
+	k_sem_take(&priv->sem, K_FOREVER);
+
+#ifdef STM32_SDMMC_USE_DMA
```
Doesn't seem like a correct flag

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -139,15 +180,43 @@ static int stm32_sdmmc_access_read(struct disk_info *disk, uint8_t *data_buf,
-		;
+	k_sem_take(&priv->sync, K_FOREVER);
+
+#ifdef STM32_SDMMC_USE_DMA
```
Wrong flag

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -160,14 +229,35 @@ static int stm32_sdmmc_access_write(struct disk_info *disk,
-				 num_sector, 30000);
+	k_sem_take(&priv->sem, K_FOREVER);
+
+#ifdef STM32_SDMMC_USE_DMA
```
Wrong flag

erwango commented on subsys/disk/Kconfig:
```diff
@@ -144,11 +144,37 @@ config DISK_ACCESS_STM32_SDMMC
 	select USE_STM32_HAL_SD
+	select USE_STM32_HAL_DMA if SOC_SERIES_STM32F7X
 	select USE_STM32_LL_SDMMC
+	select USE_STM32_LL_DELAYBLOCK
```
if  SOC_SERIES_STM32H7X

erwango reviewed (changes requested):
Few  changes requested.
Also, since the change is wider now, it is getting difficult to review carefully, so it would be easier to have the commit impacting the driver split in 2 parts:
-First part introducing IRQ (no H7 involved)
-Second part introducing (using H7)

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -401,41 +495,81 @@ static int disk_stm32_sdmmc_init(const struct device *dev)
-DEVICE_DT_INST_DEFINE(0, disk_stm32_sdmmc_init, device_pm_control_nop,
-		    &stm32_sdmmc_priv_1, NULL, APPLICATION,
-		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    NULL);
-#endif
+#define STM32_SDMMC_INIT(id)						\
+									\
+static void sdmmc_irqconfig_##id(const struct device *dev)		\
+{									\
+	IRQ_CONNECT(DT_INST_IRQN(id), DT_INST_IRQ(id, priority),	\
+		    disk_stm32_sdmmc_isr, DEVICE_DT_INST_GET(id), 0);	\
+	irq_enable(DT_INST_IRQN(0));					\
+}									\
+									\
+static const struct soc_gpio_pinctrl sdmmc_pins_##id[] =		\
+	ST_STM32_DT_INST_PINCTRL(id, 0);				\
+									\
+static struct stm32_sdmmc_priv stm32_sdmmc_priv_##id = {		\
+	.hsd = {							\
+		.Instance = (SDMMC_TypeDef *)DT_INST_REG_ADDR(id),	\
+	},								\
+	.pclken = {							\
+		.bus = DT_INST_CLOCKS_CELL(id, bus),			\
+		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
+	},								\
+	.pinctrl = {							\
+		.list = sdmmc_pins_##id,				\
+		.len = ARRAY_SIZE(sdmmc_pins_##id)			\
+	},								\
+	.irq_config = sdmmc_irqconfig_##id,				\
+	STM32_SDMMC_GPIO(id, cd),					\
+	STM32_SDMMC_GPIO(id, pe),					\
+};									\
+									\
+DEVICE_DT_INST_DEFINE(id, disk_stm32_sdmmc_init, device_pm_control_nop,	\
+		    &stm32_sdmmc_priv_##id, NULL, APPLICATION,		\
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
+		    NULL)
+
+DT_INST_FOREACH_STATUS_OKAY(STM32_SDMMC_INIT);
```
@affrinpinhero-2356 > Multi-instance support
